staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model/schemas.go

```diff
@@ -40,152 +33,7 @@ const maxRequestSizeBytes = apiservercel.DefaultMaxRequestSizeBytes
-				// properties
-				// only include required properties without a default value; default values are filled in
-				// server-side
-				if required[name] && prop.Default.Object == nil {
-					minSerializedSize += int64(len(name)) + fieldType.MinSerializedSize + 4
-				}
-			}
-		}
-		objType := apiservercel.NewObjectType("object", fields)
-		objType.MinSerializedSize = minSerializedSize
-		return objType
-	case "string":
-		if s.ValueValidation != nil {
-			switch s.ValueValidation.Format {
-			case "byte":
-				byteWithMaxLength := apiservercel.NewSimpleTypeWithMinSize("bytes", cel.BytesType, types.Bytes([]byte{}), apiservercel.MinStringSize)
-				if s.ValueValidation.MaxLength != nil {
-					byteWithMaxLength.MaxElements = zeroIfNegative(*s.ValueValidation.MaxLength)
-				} else {
-					byteWithMaxLength.MaxElements = estimateMaxStringLengthPerRequest(s)
-				}
-				return byteWithMaxLength
-			case "duration":
-				durationWithMaxLength := apiservercel.NewSimpleTypeWithMinSize("duration", cel.DurationType, types.Duration{Duration: time.Duration(0)}, int64(apiservercel.MinDurationSizeJSON))
-				durationWithMaxLength.MaxElements = estimateMaxStringLengthPerRequest(s)
-				return durationWithMaxLength
-			case "date":
-				timestampWithMaxLength := apiservercel.NewSimpleTypeWithMinSize("timestamp", cel.TimestampType, types.Timestamp{Time: time.Time{}}, int64(apiservercel.JSONDateSize))
-				timestampWithMaxLength.MaxElements = estimateMaxStringLengthPerRequest(s)
-				return timestampWithMaxLength
-			case "date-time":
-				timestampWithMaxLength := apiservercel.NewSimpleTypeWithMinSize("timestamp", cel.TimestampType, types.Timestamp{Time: time.Time{}}, int64(apiservercel.MinDatetimeSizeJSON))
-				timestampWithMaxLength.MaxElements = estimateMaxStringLengthPerRequest(s)
-				return timestampWithMaxLength
-			}
-		}
-		strWithMaxLength := apiservercel.NewSimpleTypeWithMinSize("string", cel.StringType, types.String(""), apiservercel.MinStringSize)
-		if s.ValueValidation != nil && s.ValueValidation.MaxLength != nil {
-			// multiply the user-provided max length by 4 in the case of an otherwise-untyped string
-			// we do this because the OpenAPIv3 spec indicates that maxLength is specified in runes/code points,
-			// but we need to reason about length for things like request size, so we use bytes in this code (and an individual
-			// unicode code point can be up to 4 bytes long)
-			strWithMaxLength.MaxElements = zeroIfNegative(*s.ValueValidation.MaxLength) * 4
-		} else {
-			strWithMaxLength.MaxElements = estimateMaxStringLengthPerRequest(s)
-		}
-		return strWithMaxLength
-	case "boolean":
-		return apiservercel.BoolType
-	case "number":
-		return apiservercel.DoubleType
-	case "integer":
-		return apiservercel.IntType
-	}
-	return nil
-}
-
-func zeroIfNegative(v int64) int64 {
-	if v < 0 {
-		return 0
-	}
-	return v
+	return celopenapi.SchemaDeclType(s.ToKubeOpenAPI(), isResourceRoot)
```

Comment:
same question about whether ToKubeOpenAPI() is called once per validator instance or once per validation

Reply:
it is called once per validation, every time the variables like `self`, `oldSelf` is constructed.
I've updated the solution to avoid the conversion with the adapter.